core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1499,30 +1500,54 @@ class KafkaApis(val requestChannel: RequestChannel,
   def handleLeaveGroupRequest(request: RequestChannel.Request) {
     val leaveGroupRequest = request.body[LeaveGroupRequest]
 
-    // the callback for sending a leave-group response
-    def sendResponseCallback(error: Errors) {
-      def createResponse(requestThrottleMs: Int): AbstractResponse = {
-        val response = new LeaveGroupResponse(new LeaveGroupResponseData()
-          .setThrottleTimeMs(requestThrottleMs)
-          .setErrorCode(error.code()))
-        trace("Sending leave group response %s for correlation id %d to client %s."
-          .format(response, request.header.correlationId, request.header.clientId))
-        response
-      }
-      sendResponseMaybeThrottle(request, createResponse)
-    }
-
     if (!authorize(request.session, Read, Resource(Group, leaveGroupRequest.data().groupId(), LITERAL))) {
       sendResponseMaybeThrottle(request, requestThrottleMs =>
         new LeaveGroupResponse(new LeaveGroupResponseData()
           .setThrottleTimeMs(requestThrottleMs)
-          .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code())))
+            .setMemberResponseList(Collections.singletonList(
+              new MemberResponse()
+                .setErrorCode(Errors.GROUP_AUTHORIZATION_FAILED.code)
+            ))
+        )
+      )
     } else {
-      // let the coordinator to handle leave-group
-      groupCoordinator.handleLeaveGroup(
-        leaveGroupRequest.data().groupId(),
-        leaveGroupRequest.data().memberId(),
-        sendResponseCallback)
+      val leavingMembers = leaveGroupRequest.data.memberIdentityList
+      val memberResponseList = new util.ArrayList[LeaveGroupResponseData.MemberResponse]()
+      val totalMembers = leavingMembers.size
+
+      def sendResponseCallback(leaveGroupResult: LeaveGroupResult) {
+        memberResponseList.add(
+          new MemberResponse()
+            .setMemberId(leaveGroupResult.memberId)
+            .setGroupInstanceId(leaveGroupResult.groupInstanceId.orNull)
+            .setErrorCode(leaveGroupResult.error.code)
+        )
+
+        // Send response only after we have gone through all the leaving members.
+        if (totalMembers == memberResponseList.size) {
+          def createResponse(requestThrottleMs: Int): AbstractResponse = {
+            val batchLeaveGroupResponse = new LeaveGroupResponse(
+              new LeaveGroupResponseData()
+                .setThrottleTimeMs(requestThrottleMs)
+                .setMemberResponseList(memberResponseList)
+            )
+
+            trace("Sending leave group response %s for correlation id %d to client %s."
+              .format(batchLeaveGroupResponse, request.header.correlationId, request.header.clientId))
+            batchLeaveGroupResponse
+          }
+          sendResponseMaybeThrottle(request, createResponse)
+        }
+      }
+
+      for (member <- leavingMembers.asScala) {
```

Comment:
This is a meta question: instead of making the callee to do one member at a time, could we just refactor it to make the callee taking a list of (memberId, instanceId) instead, so that we can avoid some race condition that in between `group.inLock` some other thread may kicks in and modified the group, i.e. some members are successfully removed while others are not? I think technically speaking it is allowed since we return one error code per member anyways, but client-side handling different error codes for members that some succeeded and some failed could be clumsy.

Reply:
@hachikuji